stdlib/contracts.ncl

```diff
@@ -73,7 +73,80 @@
       else %blame% (%tag% "extra field `#{%head% (%fieldsOf% t)}`" l),
 
   contracts = {
-    blame = fun l => %blame% l,
-    tag = fun msg l => %tag% msg l,
+    blame
+      | doc m#"
+        Raise blame for a given label.
+
+        Type: `forall a. Lbl -> a`
+        (for technical reasons, this element isn't actually statically typed)
+
+        Blame is the mechanism to signal contract violiation in Nickel. It ends
+        the program execution and print a detailed report thanks to the
+        information tracked inside the label.
+
+        For example:
+        ```nickel
+        IsZero = fun label value =>
+          if value == 0 then value
+          else contracts.blame label
+        ```
+        "#m
+      = fun l => %blame% l,
+    blameWith
+      | doc m#"
+        Raise blame with respect to a given label and a custom error message.
+
+        Type: `forall a. Str -> Lbl -> a`
+        (for technical reasons, this element isn't actually statically typed)
+
+        Same as `blame`, but take an additional custom error message that will be
+        displayed as part of the blame error. `blameWith msg l` is equivalent to
+        `blame (tag msg l)
+
+        For example:
+        ```nickel
+        let IsZero = fun label value =>
+          if value == 0 then value
+          else contracts.blame label in
+        0 | #IsZero
+        ```
+        "#m
+      = fun msg l => %blame% (%tag% msg l),
+    fromPred
+      | doc m#"
+        Generate a contract from a boolean predicate.
+
+        Type: `(Dyn -> Bool) -> (Lbl -> Dyn -> Dyn)`
+        (for technical reasons, this element isn't actually statically typed)
+
+        For example:
+        ```
+        let IsZero = contracts.fromPred (fun x => x == 0) in
+        0 | #IsZero
+        ```
+        "#m
+      = fun pred l v => if pred v then v else %blame% l,
```

Comment:
Should this also take a tag as argument to encourage writing meaningful error messages?

```
let IsZero = contracts.fromPred "Is not zero" (fun x => x == 0) in
...
```

Reply:
I wondered the same thing. But note that a predicate can only fail in one way (if it evaluates to false). Because blame errors report the original contract annotation, I think in most case a generic error message like "argument is non zero" wouldn't add much to, say, just the name of the contract `NonZero`. Or at least that probably we shouldn't force users to attach such a message when the information that the contract `NonZero` or `GreaterThan 10` failed is sufficient.

However, maybe having a variant `fromPredWith` (or whatever we call it) is a good idea. Would you like me to add it to this PR?

Reply:
Oh yes, the name appears in the report, ofc.

> However, maybe having a variant fromPredWith (or whatever we call it) is a good idea. Would you like me to add it to this PR?

You’ve just convinced me that it wasn’t necessary it shouldn’t be necessary in most cases :stuck_out_tongue: . So I’d say that we can keep it out for now and see whether `tag "foo" (fromPred p)` becomes a commonly needed pattern